flink-runtime/src/main/java/org/apache/flink/runtime/state/TaskLocalStateStoreImpl.java

```diff
@@ -141,6 +158,29 @@ public void storeLocalState(
                     subtaskIndex);
         }
 
+        for (Map.Entry<OperatorID, OperatorSubtaskState> subtaskStateEntry :
+                localState.getSubtaskStateMappings()) {
+            for (KeyedStateHandle keyedStateHandle :
+                    subtaskStateEntry.getValue().getManagedKeyedState()) {
+                if (keyedStateHandle instanceof ChangelogStateBackendHandle) {
+                    ChangelogStateBackendHandle changelogStateBackendHandle =
+                            (ChangelogStateBackendHandle) keyedStateHandle;
+                    long materializationID = changelogStateBackendHandle.getMaterializationID();
+                    if (currentMaterializationID == null
+                            || materializationID != currentMaterializationID.f0) {
+                        currentMaterializationID = Tuple2.of(materializationID, checkpointId);
+                        referredByCheckpointID.clear();
+                    } else {
+                        referredByCheckpointID.put(currentMaterializationID.f1, checkpointId);
```

Comment:
Shouldn't this (and other access) be synchronized, similar to `storedTaskStateByCheckpointID`?

The method is called from `AsyncCheckpointRunnable`, which can be executed by different threads (plus other methods).

Reply:
Yes, move it into `synchronized` block.